ruby/ql/src/experimental/decompression-api/DecompressionApi.ql

```diff
@@ -0,0 +1,56 @@
+/**
+ * @name User-controlled file decompression
+ * @description User-controlled data that flows into decompression library APIs could be dangerous
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 7.8
+ * @precision medium
+ * @id rb/user-controlled-file-decompression
+ * @tags security external/cwe/cwe-409
+ */
+
+import ruby
+import codeql.ruby.ApiGraphs
+import codeql.ruby.DataFlow
+import codeql.ruby.dataflow.RemoteFlowSources
+import codeql.ruby.dataflow.BarrierGuards
+import codeql.ruby.TaintTracking
+import DataFlow::PathGraph
+
+class DecompressionApiUse extends DataFlow::Node {
+
+    // this should find the first argument of Zlib::Inflate.inflate or Zip::File.extract
+    DecompressionApiUse() {
+        this = API::getTopLevelMember("Zlib").getMember("Inflate").getAMethodCall("inflate").getArgument(0) or
+        this = API::getTopLevelMember("Zip").getMember("Entry").getAMethodCall("extract").getArgument(0)
```

Comment:
Yep, you're right. I'm struggling to think of a good way to query for this is terms of tainted data flowing to `.extract` if we're not constraining to an argument. From what I can find, it looks like [this method](https://www.rubydoc.info/gems/rubyzip/1.2.0/Zip/CentralDirectory#read_from_stream-instance_method) is called under the hood in `Zip::File.open`, so would it make sense to query for tainted data that flows to that or `Zip::File.open`? Or does that seem like I'm going about this the wrong way and should just alert for any usage of `Zip::Entry.extract` since this query is shaping up to be pretty noisy anyway?

Reply:
treating `Zip::File.open` and `Zip::File.new` as a sinks makes sense. We can add some barrier guard to exclude flow of string values (file paths). Those should be caught the path traversal query. In addition you probably want to add `Zip::File.open_buffer`; for this sink there should be no barrier guard, because a string value is treated as contents not as file path.

Reply:
Tracking `Zip::Entry.extract` instead would avoid flagging up cases where a zip file is opened but no data is extracted. However, opening a zip file and not extracting anything from it is likely very rare in practice. In addition there are many different ways data can be extracted from a `Zip::Entry`, so tracking `Zip::File.open` should be simpler. 

We might want to augment the query later on with some additional checks to reduce false positives. For example if the code has tests that `Zip::Entry.size` is within reasonable bounds or the compression rate is "sane" then that mitigates the effect of a zip bomb.

Reply:
Ok, this makes sense. I'm going to wait until https://github.com/github/codeql/pull/9477 is finished and merged and then I'll model `Zip::File.open_buffer` in Archives.qll and add it to this query. I've also dropped the barrier guards here since they don't make sense in this context.